From a385856b59c8289cc7273fa3a3062bbf0ef96c97 Mon Sep 17 00:00:00 2001
From: Deepesh Pathak <deepesh.pathak@isovalent.com>
Date: Sun, 26 Oct 2025 13:27:15 -0700
Subject: [PATCH] policy: fix {to,from}-groups derived policy creation for
 empty cidr sets

This commit fixes a day-0 bug in cilium-operator when processing to-groups
derived policy resolving to an empty set of ip addresses(eg. non existing
security group).
Currently if the to-groups reference in a CNP/CCNP resolves to an empty
set of ip address, the resulting derived policy is created with no L3
selectors(resulting in allow-all at L3). This happens due to a bug in
marshalling {To,From}CIDRSet fields which are marked with `omitempty`
tag. This tag is inconsistent with the expected behavior of this field.

```go
type CIDRRuleSlice []CIDRRule
type EgressCommonRule struct {
    ...
    // +kubebuilder:validation:Optional
    ToCIDRSet CIDRRuleSlice `json:"toCIDRSet,omitempty"`
}

// Marshalled object: `{}`
// Intent: No L3 Selector(allow all)
// Actual: No L3 Selector
EgressCommonRule{
    ToCIDRSet: nil,
}

// Marshalled object: `{}`
// Intent: L3 Selector with no configured CIDRs(allow none)
// Actual: No L3 Selector
EgressCommonRule{
    ToCIDRSet: []CIDRRule{},
}
```

This commit changes this behavior by changing the field tags to
`omitzero` which correctly marshals nil and empty slices to
corresponding user intent.

Signed-off-by: Deepesh Pathak <deepesh.pathak@isovalent.com>
---
 pkg/policy/api/egress.go       |  6 ++--
 pkg/policy/api/egress_test.go  | 53 +++++++++++++++++++++++++++++++++-
 pkg/policy/api/ingress.go      |  6 ++--
 pkg/policy/api/ingress_test.go | 47 ++++++++++++++++++++++++++++++
 4 files changed, 105 insertions(+), 7 deletions(-)

diff --git a/pkg/policy/api/egress.go b/pkg/policy/api/egress.go
index 284f35d137ab6..bcc32c75501d9 100644
--- a/pkg/policy/api/egress.go
+++ b/pkg/policy/api/egress.go
@@ -54,7 +54,7 @@ type EgressCommonRule struct {
 	// initiate connections to 10.2.3.0/24 except from IPs in subnet 10.2.3.0/28.
 	//
 	// +kubebuilder:validation:Optional
-	ToCIDRSet CIDRRuleSlice `json:"toCIDRSet,omitempty"`
+	ToCIDRSet CIDRRuleSlice `json:"toCIDRSet,omitzero"`
 
 	// ToEntities is a list of special entities to which the endpoint subject
 	// to the rule is allowed to initiate connections. Supported entities are
@@ -243,7 +243,7 @@ func (e *EgressRule) CreateDerivative(ctx context.Context) (*EgressRule, error)
 	if err != nil {
 		return &EgressRule{}, err
 	}
-	newRule.ToCIDRSet = append(e.ToCIDRSet, cidrSet...)
+	newRule.ToCIDRSet = append(newRule.ToCIDRSet, cidrSet...)
 	newRule.ToGroups = nil
 	return newRule, nil
 }
@@ -262,7 +262,7 @@ func (e *EgressDenyRule) CreateDerivative(ctx context.Context) (*EgressDenyRule,
 	if err != nil {
 		return &EgressDenyRule{}, err
 	}
-	newRule.ToCIDRSet = append(e.ToCIDRSet, cidrSet...)
+	newRule.ToCIDRSet = append(newRule.ToCIDRSet, cidrSet...)
 	newRule.ToGroups = nil
 	return newRule, nil
 }
diff --git a/pkg/policy/api/egress_test.go b/pkg/policy/api/egress_test.go
index 9983950447e1a..0f59c2584e28c 100644
--- a/pkg/policy/api/egress_test.go
+++ b/pkg/policy/api/egress_test.go
@@ -5,6 +5,7 @@ package api
 
 import (
 	"context"
+	"encoding/json"
 	"fmt"
 	"net/netip"
 	"testing"
@@ -102,7 +103,11 @@ func TestCreateDerivativeWithoutErrorAndNoIPs(t *testing.T) {
 
 	newRule, err := eg.CreateDerivative(context.TODO())
 	require.NoError(t, err)
-	require.Equal(t, &EgressRule{}, newRule)
+	require.Equal(t, &EgressRule{
+		EgressCommonRule: EgressCommonRule{
+			ToCIDRSet: CIDRRuleSlice{},
+		},
+	}, newRule)
 }
 
 func TestEgressCommonRuleDeepEqual(t *testing.T) {
@@ -220,3 +225,49 @@ func TestEgressCommonRuleDeepEqual(t *testing.T) {
 		})
 	}
 }
+
+func TestEgressCommonRuleMarshalling(t *testing.T) {
+	testCases := []struct {
+		name     string
+		in       *EgressCommonRule
+		expected string
+	}{
+		{
+			name: "ToCIDRSet is nil",
+			in: &EgressCommonRule{
+				ToCIDRSet: nil,
+			},
+			expected: `{}`,
+		},
+		{
+			name: "ToCIDRSet is empty",
+			in: &EgressCommonRule{
+				ToCIDRSet: []CIDRRule{},
+			},
+			expected: `{"toCIDRSet":[]}`,
+		},
+		{
+			name: "ToCIDRSet has CIDR",
+			in: &EgressCommonRule{
+				ToCIDRSet: []CIDRRule{
+					{
+						Cidr: "192.168.1.0/24",
+					},
+				},
+			},
+			expected: `{"toCIDRSet":[{"cidr":"192.168.1.0/24"}]}`,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			data, err := json.Marshal(tc.in)
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, string(data))
+
+			rule := EgressCommonRule{}
+			err = json.Unmarshal(data, &rule)
+			require.NoError(t, err)
+			require.True(t, tc.in.DeepEqual(&rule))
+		})
+	}
+}
diff --git a/pkg/policy/api/ingress.go b/pkg/policy/api/ingress.go
index fadbc69c127f8..9b21c4331d37f 100644
--- a/pkg/policy/api/ingress.go
+++ b/pkg/policy/api/ingress.go
@@ -55,7 +55,7 @@ type IngressCommonRule struct {
 	// connections from 10.0.0.0/8 except from IPs in subnet 10.96.0.0/12.
 	//
 	// +kubebuilder:validation:Optional
-	FromCIDRSet CIDRRuleSlice `json:"fromCIDRSet,omitempty"`
+	FromCIDRSet CIDRRuleSlice `json:"fromCIDRSet,omitzero"`
 
 	// FromEntities is a list of special entities which the endpoint subject
 	// to the rule is allowed to receive connections from. Supported entities are
@@ -225,7 +225,7 @@ func (e *IngressRule) CreateDerivative(ctx context.Context) (*IngressRule, error
 	if err != nil {
 		return &IngressRule{}, err
 	}
-	newRule.FromCIDRSet = append(e.FromCIDRSet, cidrSet...)
+	newRule.FromCIDRSet = append(newRule.FromCIDRSet, cidrSet...)
 	newRule.FromGroups = nil
 	return newRule, nil
 }
@@ -244,7 +244,7 @@ func (e *IngressDenyRule) CreateDerivative(ctx context.Context) (*IngressDenyRul
 	if err != nil {
 		return &IngressDenyRule{}, err
 	}
-	newRule.FromCIDRSet = append(e.FromCIDRSet, cidrSet...)
+	newRule.FromCIDRSet = append(newRule.FromCIDRSet, cidrSet...)
 	newRule.FromGroups = nil
 	return newRule, nil
 }
diff --git a/pkg/policy/api/ingress_test.go b/pkg/policy/api/ingress_test.go
index 5e3cb55f36238..c9b146d942cf2 100644
--- a/pkg/policy/api/ingress_test.go
+++ b/pkg/policy/api/ingress_test.go
@@ -5,6 +5,7 @@ package api
 
 import (
 	"context"
+	"encoding/json"
 	"testing"
 
 	"github.com/stretchr/testify/require"
@@ -179,3 +180,49 @@ func TestIngressCommonRuleDeepEqual(t *testing.T) {
 		})
 	}
 }
+
+func TestIngressCommonRuleMarshalling(t *testing.T) {
+	testCases := []struct {
+		name     string
+		in       *IngressCommonRule
+		expected string
+	}{
+		{
+			name: "ToCIDRSet is nil",
+			in: &IngressCommonRule{
+				FromCIDRSet: nil,
+			},
+			expected: `{}`,
+		},
+		{
+			name: "ToCIDRSet is empty",
+			in: &IngressCommonRule{
+				FromCIDRSet: []CIDRRule{},
+			},
+			expected: `{"fromCIDRSet":[]}`,
+		},
+		{
+			name: "ToCIDRSet has CIDR",
+			in: &IngressCommonRule{
+				FromCIDRSet: []CIDRRule{
+					{
+						Cidr: "192.168.1.0/24",
+					},
+				},
+			},
+			expected: `{"fromCIDRSet":[{"cidr":"192.168.1.0/24"}]}`,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			data, err := json.Marshal(tc.in)
+			require.NoError(t, err)
+			require.Equal(t, tc.expected, string(data))
+
+			rule := IngressCommonRule{}
+			err = json.Unmarshal(data, &rule)
+			require.NoError(t, err)
+			require.True(t, tc.in.DeepEqual(&rule))
+		})
+	}
+}
